### PR TITLE
[Core] Adding a few methods to MathUtils

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
Adding two required methods for #8903 :

- Calculate the exponential of a matrix
- Factorial of an integer

### CODE REVIEW

File: kratos/tests/cpp_tests/utilities/test_math_utils.cpp [hunk 1]
```diff
@@ -870,5 +870,30 @@ namespace Kratos
...
+        KRATOS_TEST_CASE_IN_SUITE(MathUtilsExponentialOfMatrix, KratosCoreFastSuite)
+        {
+            Matrix A(3, 3), exp_A(3, 3);
+            noalias(A) = IdentityMatrix(3);
+            A(1, 1) = 2.0;
+            A(2, 2) = 3.0;
+            MathUtils<double>::CalculateExponentialOfMatrix(A, exp_A);
+            KRATOS_CHECK_NEAR(exp_A(0, 0), std::exp(A(0, 0)), 1.0e-4);
+            KRATOS_CHECK_NEAR(exp_A(1, 1), std::exp(A(1, 1)), 1.0e-4);
+            KRATOS_CHECK_NEAR(exp_A(2, 2), std::exp(A(2, 2)), 1.0e-4);
```
Comment (philbucher) [inline kratos/tests/cpp_tests/utilities/test_math_utils.cpp]: can you please use the matrix check macro?
And why such low tolerance?

Comment (AlejandroCornejo) [reply kratos/tests/cpp_tests/utilities/test_math_utils.cpp]: done it, BTW, How do you run the cpp tests of the CORE????

File: kratos/utilities/math_utils.h [hunk 2]
```diff
@@ -1876,6 +1876,56 @@ class MathUtils
...
     }
 
+    /**
+     * @brief Calculates the Factorial of a number k, Factorial = k!
+     * @tparam Number The number of which the Factorial is computed
+     */
+    template<class TIntegerType>
+    static inline TIntegerType Factorial(
+        const TIntegerType Number
+        )
```
Comment (rubenzorrilla) [inline kratos/utilities/math_utils.h]:
```suggestion
    static inline TIntegerType Factorial(const TIntegerType Number)
```
Style guide for one argument functions.

File: kratos/utilities/math_utils.h [hunk 3]
```diff
@@ -1876,6 +1876,56 @@ class MathUtils
...
+    {
+        if (Number == 0) {
+            return 1;
+        }
+        TIntegerType k = Number;
+        for (unsigned int i = Number - 1; i > 0; --i){
```
Comment (rubenzorrilla) [inline kratos/utilities/math_utils.h]:
```suggestion
        for (TIntegerType i = Number - 1; i > 0; --i){
```

File: kratos/tests/cpp_tests/utilities/test_math_utils.cpp (hunk 1 above)
Comment (philbucher) [reply kratos/tests/cpp_tests/utilities/test_math_utils.cpp]: `python kratos/python_scripts/run_cpp_tests.py`

File: kratos/utilities/math_utils.h [hunk 4]
```diff
@@ -1876,6 +1876,56 @@ class MathUtils
...
+     * @tparam rMatrix: the matrix A of which exp is calculated
+     * @tparam rExponentialMatrix: exp(A)
+     */
+    template<class TMatrixType>
+    static inline void CalculateExponentialOfMatrix(
+            const TMatrixType& rMatrix,
+            TMatrixType& rExponentialMatrix
+        )
+    {
+        int series_term = 2, max_terms = 200, factorial = 1;
```
Comment (rubenzorrilla) [inline kratos/utilities/math_utils.h]:
```suggestion
        std::size_t series_term = 2;
        std::size_t factorial = 1;
        const std::size_t max_terms = 200;
```
Shouldn't we expose the `max_terms` to the user (defaulting it to the current value)?

File: kratos/utilities/math_utils.h [hunk 5]
```diff
@@ -1876,6 +1876,56 @@ class MathUtils
...
+        TMatrixType exponent_matrix = rMatrix;
+        TMatrixType aux_matrix;
+
+        while (series_term < max_terms) {
+            noalias(aux_matrix) = prod(exponent_matrix, rMatrix);
+            noalias(exponent_matrix) = aux_matrix;
+            factorial = Factorial(series_term);
+            noalias(rExponentialMatrix) += exponent_matrix / factorial;
+            const double norm_series_term = std::abs(norm_frobenius(exponent_matrix) / factorial);
+            if (norm_series_term < 1000.0*ZeroTolerance)
```
Comment (rubenzorrilla) [inline kratos/utilities/math_utils.h]: Shouldn't this tolerance be exposed?

Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: I agree

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Tolerance should be an argument as @rubenzorrilla says

File: kratos/utilities/math_utils.h (hunk 5 above)
Comment (AlejandroCornejo) [reply kratos/utilities/math_utils.h]: done

**Comment by AlejandroCornejo** [comment]
DONE! let's see if the checks are ok

**Review by loumalouomega — DISMISSED** [review]
Now is OK

File: kratos/utilities/math_utils.h [hunk 6]
```diff
@@ -1876,6 +1876,57 @@ class MathUtils
...
+
+    /**
+     * @brief Calculates the exponential of a matrix
+     * @brief see https://mathworld.wolfram.com/MatrixExponential.html
+     * @tparam rMatrix: the matrix A of which exp is calculated
+     * @tparam rExponentialMatrix: exp(A)
+     */
+    template<class TMatrixType>
+    static inline void CalculateExponentialOfMatrix(
+            const TMatrixType& rMatrix,
```
Comment (loumalouomega) [inline kratos/utilities/math_utils.h]: Minor comments, so many spaces?

**Review by loumalouomega — APPROVED** [review]
ReOK

**Pull request merged**